components/org.wso2.carbon.identity.application.authenticator.push.device.handler/src/main/java/org/wso2/carbon/identity/application/authenticator/push/device/handler/dao/DeviceDAOImpl.java

```diff
@@ -0,0 +1,217 @@
+            if (!connection.getAutoCommit()) {
+                connection.commit();
+            }
+        } finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, preparedStatement);
+        }
+    }
+
+    @Override
+    public void unregisterDevice(String deviceId) throws SQLException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        PreparedStatement preparedStatement = null;
+        try {
+            preparedStatement = connection.prepareStatement(DeviceHandlerConstants.SQLQueries.UNREGISTER_DEVICE);
+            preparedStatement.setString(1, deviceId);
+            preparedStatement.execute();
+        } finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, preparedStatement);
+        }
+    }
+
+    @Override
+    public void editDevice(String deviceId, Device updatedDevice) throws SQLException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        PreparedStatement preparedStatement = null;
+        try {
+            preparedStatement = connection.prepareStatement(DeviceHandlerConstants.SQLQueries.EDIT_DEVICE);
+            preparedStatement.setString(1, updatedDevice.getDeviceName());
+            preparedStatement.setString(2, updatedDevice.getPushId());
+            preparedStatement.setString(3, deviceId);
+            preparedStatement.execute();
+        } finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, preparedStatement);
+        }
+
+    }
+
+    @Override
+    public Device getDevice(String deviceId) throws PushDeviceHandlerServerException, SQLException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        PreparedStatement preparedStatement = null;
+        Device device = new Device();
+        ResultSet resultSet = null;
+        try {
+            preparedStatement = connection.prepareStatement(DeviceHandlerConstants.SQLQueries.GET_DEVICE);
+            preparedStatement.setString(1, deviceId);
+            resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                device.setDeviceId(resultSet.getString(1));
+                device.setDeviceName(resultSet.getString(2));
+                device.setDeviceModel(resultSet.getString(3));
+                device.setPushId(resultSet.getString(4));
+                device.setPublicKey(resultSet.getString(5));
+                device.setRegistrationTime(timestampToDate(resultSet.getTimestamp(6)));
+                device.setLastUsedTime(timestampToDate(resultSet.getTimestamp(7)));
+            } else {
+                IdentityDatabaseUtil.closeAllConnections(connection, resultSet, preparedStatement);
+                String errorMessage =
+                        String.format("The requested device: %s is not registered in the system.", deviceId);
+                throw new PushDeviceHandlerServerException(errorMessage);
```

Comment:
should be client exception

Reply:
Return optional to handle null instead. Throw exception at service